### PR TITLE
Moved creating SiPixelTemplateStore from DB to an ESProducer [13_2]

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

All DB construction was switched to using EventSetup. Cases using reading from files were kept local to module.
The new ESProducer was added as a Task to all needed Sequences.

#### PR validation:

Code compiles. 

backport of #42514 and #42597

### CODE REVIEW

**Comment by mmusich** [comment]
this branch has conflicts to be resolved.

**Comment by missirol** [comment]
+hlt

(This update will require the 'parsing' of the next CMSSW_13_2_X release into ConfDB, in order to add the new ESProducer to the HLT menus in the db, I think.)

**Comment by missirol** [comment]
backport of https://github.com/cms-sw/cmssw/pull/42514
backport of https://github.com/cms-sw/cmssw/pull/42597

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
 from RecoTracker.Configuration.RecoTrackerP5_cff import *
 from RecoTracker.TrackProducer.TrackRefitter_cfi import *
 from TrackingTools.TransientTrack.TransientTrackBuilder_cfi import *
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import *
```
Comment (perrotta) [inline Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py:29]: I should have had commented already in #42597...

But as of now these `include`'s in the modified Calibration/TkAlCaRecoProducers/python configs are not needed any more. Should they get removed, or do you plan to let them there as a placeholder for future reinsertion of Tasks in them?

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 2]
```diff
@@ -215,12 +215,20 @@ def customiseForOffline(process):
     return process
 
 
+def customizeHLTfor42514(process):
```
Comment (perrotta) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py:218]: @missirol @Martin-Grunewald I don't remember: should we put here the number of this PR, or the one in the master (as  implemented here)?

**Review by perrotta — COMMENTED** [review]
The three backports in 13_2_X, 13_1_X and 13_0_X are identical, and all verbatim backports of the combination of #42514 and #42597. Therefore the same comments made here are also valid for #42647 and #42648

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py (hunk 2 above)
Comment (Martin-Grunewald) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py:218]: In principle the one for the specific release...

Comment (missirol) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py:218]: > In principle the one for the specific release...

True, but I would suggest to keep the number of the original PR. The reason is that keeping the files identical (as much as possible) across cycles helps when cherry-picking, avoiding silly conflicts.

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py (hunk 1 above)
Comment (Dr15Jones) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py:29]: I don't have any (short term) plans on reinserting Tasks.

Comment (perrotta) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py:29]: > I don't have any (short term) plans on reinserting Tasks.

Thank you @Dr15Jones 
Should we then clean up these backports from those unnecessary includes, together with the master (with an additional PR)?

Comment (mmusich) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py:29]: yes - please.

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngle_cff.py [hunk 3]
```diff
@@ -62,9 +62,10 @@
                                                    )
 
 # The actual sequence
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import *
```
Comment (mmusich) [inline Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngle_cff.py:65]: is it still needed here?

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripGainsAAG_cff.py [hunk 4]
```diff
@@ -63,7 +63,8 @@
                                                      )
 
 # refit and BS can be dropped if done together with RECO.
-# track filter can be moved in acalreco if no otehr users
+# track filter can be moved in acalreco if no other users
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import SiPixelTemplateStoreESProducer
```
Comment (mmusich) [inline Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripGainsAAG_cff.py:67]: here too?

File: Calibration/TkAlCaRecoProducers/python/ALCARECOSiPixelCalSingleMuonTight_cff.py [hunk 5]
```diff
@@ -76,6 +76,7 @@
 ##################################################################
 # Sequence: track refit + close-by-pixel producer
 ##################################################################
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import SiPixelTemplateStoreESProducer
```
Comment (mmusich) [inline Calibration/TkAlCaRecoProducers/python/ALCARECOSiPixelCalSingleMuonTight_cff.py:79]: and here?

**Comment by perrotta** [comment]
> Removing all the unnecessary `imports` without first testing them in the master is beyond what I'm willing to do. A domain expert should take over if a different backport is wanted.

Indeed, I'd remove them for master first, and then backport here.
I can do the master PR tomorrow (this evening is already too late for me)

**Comment by mmusich** [comment]
> Indeed, I'd remove them for master first, and then backport here.
> I can do the master PR tomorrow (this evening is already too late for me)

https://github.com/cms-sw/cmssw/pull/42656

**Comment by perrotta** [comment]
@Dr15Jones https://github.com/cms-sw/cmssw/pull/42656 had been merged in master: could you please backport it here and in the other PRs for 13_1 and 13_0?


In https://github.com/cms-sw/cmssw/pull/42656 a few includes that could be possibly removed were left untouched. Since you, Chris, added them to the original https://github.com/cms-sw/cmssw/pull/42514 PR, maybe you remember why you did it and if it is possible to remove them also. In particular:

In the following configs https://github.com/cms-sw/cmssw/pull/42514 both added the include and inserted the module SiPixelTemplateStoreESProducer in the paths, while subsequent https://github.com/cms-sw/cmssw/pull/42597 removed the module from the paths but not the include from the config:

- FastSimulation/Configuration/python/Reconstruction_BefMix_cff.py
- RecoTracker/FinalTrackSelectors/python/MergeTrackCollections_cff.py

For the following configs, instead, only the include was added by https://github.com/cms-sw/cmssw/pull/42514, but the module therein was never used. I have the impression that it was slipped there by mistake, in particular for the DQM one, but I may be wrong:

- DQMOffline/Trigger/python/SiStrip_OfflineMonitoring_cff.py

- RecoTracker/TransientTrackingRecHit/python/TTRHBuilders_cff.py

**Comment by Dr15Jones** [comment]
> maybe you remember why you did it and if it is possible to remove them also. 

So the original change was when the EventSetup module, (i.e. ES module) was assigned to a `cms.Task`. When an ES module is assigned to a cms.Task then all places where that module could be used must also be assigned to a `cms.Task` and that task must be added to the appropriate `Sequence`, `Path` or `EndPath`. So almost all of the configuration changes were to add the appropriate `cms.Task` to a `Sequence` or `Path`. In order to do that, I had to also `import` the cfi.py file, even if that cfi.py might be indirectly included by some other import. When I decided that the use of `cms.Task` was probably a bad idea at this point and removed the `cms.Task`, I failed to also remove the `import`. 

So removing the `import` can be safely done for MOST files but one still must be sure that at least one file in the full job configuration does an `import` so that the ES module is actually added to the job.

Even in the original PR, I tried to have the `import` done in the file which also imports the ES module which uses the products from SiPixelTemplateStoreESProducer. I THINK that was `TTRHBuilderWithTemplate` but that could be wrong and there could be others.

Essentially, I added the `cms.Task` based on what unit tests and RelVal tells failed. I never ran the same test after removing the `cms.Task` and trying to remove the imports.

**Comment by Dr15Jones** [comment]
> So, it is probably safe to stay with just the removals listed in https://github.com/cms-sw/cmssw/pull/42656. And at most we'll remain with some extra include here and there. Do you agree?

That's fine with me.

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py (hunk 2 above)
Comment (mmusich) [reply HLTrigger/Configuration/python/customizeHLTforCMSSW.py:218]: @missirol @Martin-Grunewald I assume from the discussion above that we are fine with keeping the old release number. Please shout if not the case.

**Comment by mandrenguyen** [comment]
+1

**Comment by mmusich** [comment]
+hlt

**Comment by perrotta** [comment]
+alca

**Comment by perrotta** [comment]
+db

**Comment by emanueleusai** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by rappoccio** [comment]
+1

**Pull request merged**